torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -411,25 +497,26 @@ def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
             grad = grad._local_tensor
         return grad
 
+    @property
+    def _sharded_local_tensor(self) -> torch.Tensor:
+        return cast(DTensor, self.sharded_param)._local_tensor
+
     def _assert_in_states(self, *states: ShardedState) -> None:
         if self.sharded_state not in states:
             _raise_assert_with_print(
                 f"Expects to be in one of {states}, not {self.sharded_state}"
             )
 
 
-# NOTE: Unsafe here refers to not checking whether the storage is already
-# allocated or freed, respectively. We should be safe to use them since we
-# explicitly manage the state transition.
-def unsafe_alloc_storage(tensor: torch.Tensor) -> None:
-    # Skip the already-allocated check and assume that `tensor` is the base
-    # tensor to save CPU overhead
-    tensor.untyped_storage().resize_(tensor.numel() * tensor.itemsize)
+def alloc_storage(tensor: torch.Tensor) -> None:
+    size = tensor.numel() * tensor.itemsize
+    if (storage := tensor.untyped_storage()).size() != size:
+        storage.resize_(size)
 
 
-def unsafe_free_storage(tensor: torch.Tensor) -> None:
-    # Skip the already-freed check to save CPU overhead
-    tensor.untyped_storage().resize_(0)
+def free_storage(tensor: torch.Tensor) -> None:
+    if (storage := tensor.untyped_storage()).size() != 0:
```

Comment:
🤔 is the bracket here correct? seems the left most one does not have a match

Reply:
It matches with the bracket right before `.size()` :) -- it is kind of confusing :/